openshift-kube-apiserver/filters/deprecatedapirequest/controller.go

```diff
@@ -0,0 +1,169 @@
+package deprecatedapirequest
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	apiv1 "github.com/openshift/api/apiserver/v1"
+	apiclientv1 "github.com/openshift/client-go/apiserver/clientset/versioned/typed/apiserver/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers"
+)
+
+// NewController returns a Controller
+func NewController(node string, config *rest.Config) Controller {
+	return &controller{
+		client:       apiclientv1.NewForConfigOrDie(config).DeprecatedAPIRequests(),
+		nodeName:     node,
+		updatePeriod: 10 * time.Second,
+		incoming:     make(chan requestRecord, 100),
+	}
+}
+
+// Controller logs deprecated api requests
+type Controller interface {
+	IsDeprecated(resource, version, group string) bool
+	LogRequest(resource, version, group, verb, user string, timestamp time.Time)
+	Start(stop <-chan struct{})
+}
+
+// controller logs deprecated api requests periodically.
+type controller struct {
+	client       apiclientv1.DeprecatedAPIRequestInterface
+	nodeName     string
+	updatePeriod time.Duration
+	incoming     chan requestRecord
+}
+
+type requestRecord struct {
+	resource  string
+	user      string
+	verb      string
+	timestamp time.Time
+}
+
+// Deprecated return true if the resource is deprecated.
+func (c *controller) IsDeprecated(resource, version, group string) bool {
+	_, ok := deprecatedApiRemovedRelease[schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resource,
+	}]
+	return ok
+}
+
+// removedRelease of a specified resource.version.group.
+func (c *controller) removedRelease(r string) string {
+	s := strings.SplitN(r, ".", 3)
+	gvr := schema.GroupVersionResource{Resource: s[0], Version: s[1]}
+	if len(s) == 3 {
+		gvr.Group = s[2]
+	}
+	return deprecatedApiRemovedRelease[gvr]
+}
+
+// LogRequest queues an api request for logging
+func (c *controller) LogRequest(resource, version, group, verb, user string, ts time.Time) {
+	r := resource + "." + version
+	if len(group) > 0 {
+		r += "." + group
+	}
+	select {
+	case c.incoming <- requestRecord{resource: r, user: user, verb: verb, timestamp: ts}:
+	}
+}
+
+// Start the controller
+func (c *controller) Start(stop <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		klog.V(1).Infof("Shutting down deprecatedapirequest controller.")
+		cancel()
+	}()
+
+	//
+	apiRequestCounts := &apiRequestCounts{resources: map[string]*resourceRequestCounts{}}
+
+	// start goroutine that adds incoming log records to the in-memory logs
+	go func() {
+		loaded := sets.NewString()
+		for {
+			select {
+			case rr := <-c.incoming:
```

Comment:
this appears to have the impact of serializing the handling of single API request on the logical lock of this channel instead of parallelizing handling from the `Log` call.